shell/browser/api/electron_api_web_frame_main.cc

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) 2020 Samuel Maddock <sam@samuelmaddock.com>.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/api/electron_api_web_frame_main.h"
+
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "base/lazy_instance.h"
+#include "base/logging.h"
+#include "content/browser/renderer_host/frame_tree_node.h"  // nogncheck
```

Comment:
can we avoid using private APIs from `//content`?

Reply:
Currently this is required for executing JS with a user gesture. https://github.com/electron/electron/blob/cf87997176986deee6c6b64d242e2ccc7871fce7/shell/browser/api/electron_api_web_frame_main.cc#L94-L99

The public methods available on a `RenderFrameHost` lack the ability to execute JS with both a callback and a user gesture.

For reference, these are the two methods to choose from:
```cpp
// This runs the JavaScript, but without restrictions. THIS IS ONLY FOR TESTS.
virtual void ExecuteJavaScriptForTests(
    const base::string16& javascript,
    JavaScriptResultCallback callback,
    int32_t world_id = ISOLATED_WORLD_ID_GLOBAL) = 0;

// This runs the JavaScript, but without restrictions. THIS IS ONLY FOR TESTS.
// Unlike the method above, this one triggers a fake user activation
// notification to test functionalities that are gated by user
// activation.
virtual void ExecuteJavaScriptWithUserGestureForTests(
    const base::string16& javascript,
    int32_t world_id = ISOLATED_WORLD_ID_GLOBAL) = 0;
```

Also, the `*ForTests` methods were used because we have precedence for using one of them here: https://github.com/electron/electron/blob/69f1731bbbaa5211aefe64f3d155b11ab7bcb327/shell/browser/ui/inspectable_web_contents.cc#L1052